blockchain/chain_sync/src/sync.rs

```diff
@@ -197,13 +198,133 @@ where
         // Persist header chain pulled from network
         self.persist_headers(&headers)?;
 
-        // TODO iterate over headers to fill out and include message, then validate message and state
-        // TODO implement sync_messages_check_state see = https://github.com/filecoin-project/lotus/blob/master/chain/sync.go#L1032
-        // TODO update SyncState::Follow
+        // Sync and validate messages from fetched tipsets
+        self.sync_messages_check_state(&headers).await?;
+        self.set_state(SyncState::Follow);
 
         Ok(())
     }
 
+    /// Syncs messages by first checking state for message existence otherwise fetches messages from blocksync
+    async fn sync_messages_check_state(&mut self, ts: &[Tipset]) -> Result<(), Error> {
+        // see https://github.com/filecoin-project/lotus/blob/master/build/params_shared.go#L109 for request window size
+        const REQUEST_WINDOW: u64 = 200;
+        // loop until i = 0;
+        loop {
+            // set i to the length of provided tipsets
+            let mut i = ts.len();
+            for tip in ts {
```

Comment:
why loop through each tipset if you are indexing and potentially reducing index by the request window? this would panic when `i` hits <0 because you are doing an unchecked subtract on an unsigned variable